packages/grid/x-data-grid/src/tests/rowSelection.DataGrid.test.tsx

```diff
@@ -774,4 +775,54 @@ describe('<DataGrid /> - Row Selection', () => {
       expect(getSelectedRowIds()).to.deep.equal([]);
     });
   });
+
+  describe('performance', () => {
+    it('should not rerender unrelated rows', () => {
+      // TODO: remove this requirement, find ways to scope down react tree rerenders.
+      const MemoizedRow = React.memo(GridRow);
+
+      // Couldn't use <RenderCounter> because we need to track multiple components
+      let commits: any[] = [];
+      function CustomCell(props: any) {
+        React.useEffect(() => {
+          commits.push({
+            rowId: props.id,
+          });
+        });
+        return <div>Hello</div>;
+      }
+
+      render(
+        <div style={{ width: 300, height: 300 }}>
+          <DataGrid
+            columns={[
+              { field: 'id', headerName: 'id', type: 'number' },
+              {
+                field: 'currencyPair',
+                headerName: 'Currency Pair',
+                renderCell: (params) => <CustomCell {...params} />,
+              },
+            ]}
+            slots={{
+              row: MemoizedRow,
+            }}
+            rows={[
+              { id: 0, currencyPair: 'USDGBP' },
+              { id: 1, currencyPair: 'USDEUR' },
+            ]}
+            autoHeight={isJSDOM}
+            checkboxSelection
+          />
+        </div>,
+      );
+      expect(getSelectedRowIds()).to.deep.equal([]);
+      expect(getRow(0).querySelector('input')).to.have.property('checked', false);
+      commits = [];
+      fireEvent.click(getCell(0, 1));
+      expect(getSelectedRowIds()).to.deep.equal([0]);
+      expect(getRow(0).querySelector('input')).to.have.property('checked', true);
+      // It shouldn't rerender rowId 1
+      expect(commits).to.deep.equal([{ rowId: 0 }]);
```

Comment:
This test runs in < 100ms (it's not super fast, but I guess it's because of the data grid). It fails in v6.4.0 (without #8942) with this error:

<img width="859" alt="Screenshot 2023-05-13 at 14 31 25" src="https://github.com/mui/mui-x/assets/3165635/2ddfd1a2-00f8-459d-9f18-538c82aefa53">